Compiler suggests dereferencing an else block

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b29e12f4b9dce6cec7c625060832ab20)):

```rust
    let a = &1;
    let b = &2;
    let val = if true {
        a + 1
    } else {
        b
    };
```

I expected to see this happen: The compiler suggests turning `b` into `*b`, since `a + 1` is an `i32` but b is a `&i32`.

Instead, this happened:

```
error[E0308]: `if` and `else` have incompatible types
  --> src/main.rs:9:9
   |
6  |       let val = if true {
   |  _______________-
7  | |         a + 1
   | |         ----- expected because of this
8  | |     } else {
9  | |         b
   | |         ^ expected integer, found `&{integer}`
10 | |     };
   | |_____- `if` and `else` have incompatible types
   |
help: consider dereferencing the borrow
   |
8  |     } else *{
9  |         b
10 |     };
   |
```

If you apply the suggestion, you get another compile error:

```
error: expected `{`, found `*`
 --> src/main.rs:6:12
  |
6 |     } else *{
  |            ^ expected `{`
  |
help: try placing this code inside a block
  |
6 |     } else { *{
7 |         b
8 |     } };
  |
```

Applying that suggestion makes it compile, but is much harder to read than just `*b`.

### Meta

`rustc --version`: 1.50.0-nightly (2020-12-03 5be3f9f10e9fd59ea038)

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"mibac138"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_typeck/src/check/demand.rs
+++ b/compiler/rustc_typeck/src/check/demand.rs
@@ -366,16 +366,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
-    crate fn hir_id_sole_block_element(
-        &self,
-        hir_id: hir::HirId,
-    ) -> Option<&'tcx rustc_hir::Expr<'tcx>> {
-        let node: Option<Node<'_>> = self.tcx.hir().find(hir_id);
-        match node {
-            Some(Node::Expr(rustc_hir::Expr {
-                kind: rustc_hir::ExprKind::Block(block, ..),
-                ..
-            })) if block.stmts.len() == 0 => block.expr,
+    /// If the given `HirId` corresponds to a block with a trailing expression, return that expression
+    crate fn maybe_get_block_expr(&self, hir_id: hir::HirId) -> Option<&'tcx hir::Expr<'tcx>> {
+        match self.tcx.hir().find(hir_id)? {
+            Node::Expr(hir::Expr { kind: hir::ExprKind::Block(block, ..), .. }) => block.expr,
             _ => None,
         }
     }
@@ -666,9 +660,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 };
                                 let suggestion = if is_struct_pat_shorthand_field {
                                     format!("{}: *{}", code, code)
-                                } else if let Some(expr) =
-                                    self.hir_id_sole_block_element(expr.hir_id)
-                                {
+                                } else if let Some(expr) = self.maybe_get_block_expr(expr.hir_id) {
                                     if let Ok(inner_code) = sm.span_to_snippet(expr.span) {
                                         format!("*{}", inner_code)
                                     } else {

--- a/compiler/rustc_typeck/src/check/demand.rs
+++ b/compiler/rustc_typeck/src/check/demand.rs
@@ -366,6 +366,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
+    crate fn hir_id_sole_block_element(
+        &self,
+        hir_id: hir::HirId,
+    ) -> Option<&'tcx rustc_hir::Expr<'tcx>> {
+        let node: Option<Node<'_>> = self.tcx.hir().find(hir_id);
+        match node {
+            Some(Node::Expr(rustc_hir::Expr {
+                kind: rustc_hir::ExprKind::Block(block, ..),
+                ..
+            })) if block.stmts.len() == 0 => block.expr,
+            _ => None,
+        }
+    }
+
     /// This function is used to determine potential "simple" improvements or users' errors and
     /// provide them useful help. For example:
     ///
@@ -652,6 +666,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 };
                                 let suggestion = if is_struct_pat_shorthand_field {
                                     format!("{}: *{}", code, code)
+                                } else if let Some(expr) =
+                                    self.hir_id_sole_block_element(expr.hir_id)
+                                {
+                                    if let Ok(inner_code) = sm.span_to_snippet(expr.span) {
+                                        format!("*{}", inner_code)
+                                    } else {
+                                        format!("*{}", code)
+                                    }
                                 } else {
                                     format!("*{}", code)
                                 };

--- a/compiler/rustc_typeck/src/check/demand.rs
+++ b/compiler/rustc_typeck/src/check/demand.rs
@@ -374,6 +374,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// Returns whether the given expression is an `else if`.
+    crate fn is_else_if_block(&self, expr: &hir::Expr<'_>) -> bool {
+        if let hir::ExprKind::If(..) = expr.kind {
+            let parent_id = self.tcx.hir().get_parent_node(expr.hir_id);
+            if let Some(Node::Expr(hir::Expr {
+                kind: hir::ExprKind::If(_, _, Some(else_expr)),
+                ..
+            })) = self.tcx.hir().find(parent_id)
+            {
+                return else_expr.hir_id == expr.hir_id;
+            }
+        }
+        false
+    }
+
     /// This function is used to determine potential "simple" improvements or users' errors and
     /// provide them useful help. For example:
     ///
@@ -660,12 +675,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 };
                                 let suggestion = if is_struct_pat_shorthand_field {
                                     format!("{}: *{}", code, code)
+                                } else if self.is_else_if_block(expr) {
+                                    // Don't suggest nonsense like `else *if`
+                                    return None;
                                 } else if let Some(expr) = self.maybe_get_block_expr(expr.hir_id) {
-                                    if let Ok(inner_code) = sm.span_to_snippet(expr.span) {
-                                        format!("*{}", inner_code)
-                                    } else {
-                                        format!("*{}", code)
-                                    }
+                                    format!("*{}", sm.span_to_snippet(expr.span).unwrap_or(code))
                                 } else {
                                     format!("*{}", code)
                                 };

--- a/src/test/ui/deref-suggestion.rs
+++ b/src/test/ui/deref-suggestion.rs
@@ -63,4 +63,12 @@ fn main() {
         b
         //~^ ERROR mismatched types
     };
+    let val = if true {
+        *a
+    } else if true {
+    //~^ ERROR incompatible types
+        b
+    } else {
+        &0
+    };
 }

--- a/src/test/ui/deref-suggestion.rs
+++ b/src/test/ui/deref-suggestion.rs
@@ -45,4 +45,14 @@ fn main() {
     //~^ ERROR mismatched types
     let r = R { i: i };
     //~^ ERROR mismatched types
+
+
+    let a = &1;
+    let b = &2;
+    let val: i32 = if true {
+        a + 1
+    } else {
+        b
+        //~^ ERROR mismatched types
+    };
 }

--- a/src/test/ui/deref-suggestion.rs
+++ b/src/test/ui/deref-suggestion.rs
@@ -55,4 +55,12 @@ fn main() {
         b
         //~^ ERROR mismatched types
     };
+    let val: i32 = if true {
+        let _ = 2;
+        a + 1
+    } else {
+        let _ = 2;
+        b
+        //~^ ERROR mismatched types
+    };
 }

--- a/src/test/ui/deref-suggestion.stderr
+++ b/src/test/ui/deref-suggestion.stderr
@@ -125,16 +125,6 @@ LL | ||     };
    | ||_____|
    | |______`if` and `else` have incompatible types
    |        expected `i32`, found `&{integer}`
-   |
-help: consider dereferencing the borrow
-   |
-LL |     } else *if true {
-LL |
-LL |         b
-LL |     } else {
-LL |         &0
-LL |     };
-   |
 
 error: aborting due to 13 previous errors
 

--- a/src/test/ui/deref-suggestion.stderr
+++ b/src/test/ui/deref-suggestion.stderr
@@ -107,6 +107,35 @@ LL |         b
    |         expected `i32`, found `&{integer}`
    |         help: consider dereferencing the borrow: `*b`
 
-error: aborting due to 12 previous errors
+error[E0308]: `if` and `else` have incompatible types
+  --> $DIR/deref-suggestion.rs:68:12
+   |
+LL |        let val = if true {
+   |   _______________-
+LL |  |         *a
+   |  |         -- expected because of this
+LL |  |     } else if true {
+   |  |____________^
+LL | ||
+LL | ||         b
+LL | ||     } else {
+LL | ||         &0
+LL | ||     };
+   | ||     ^
+   | ||_____|
+   | |______`if` and `else` have incompatible types
+   |        expected `i32`, found `&{integer}`
+   |
+help: consider dereferencing the borrow
+   |
+LL |     } else *if true {
+LL |
+LL |         b
+LL |     } else {
+LL |         &0
+LL |     };
+   |
+
+error: aborting due to 13 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/deref-suggestion.stderr
+++ b/src/test/ui/deref-suggestion.stderr
@@ -89,6 +89,15 @@ LL |     let r = R { i: i };
    |                    expected `u32`, found `&{integer}`
    |                    help: consider dereferencing the borrow: `*i`
 
-error: aborting due to 10 previous errors
+error[E0308]: mismatched types
+  --> $DIR/deref-suggestion.rs:55:9
+   |
+LL |         b
+   |         ^
+   |         |
+   |         expected `i32`, found `&{integer}`
+   |         help: consider dereferencing the borrow: `*b`
+
+error: aborting due to 11 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/deref-suggestion.stderr
+++ b/src/test/ui/deref-suggestion.stderr
@@ -98,6 +98,15 @@ LL |         b
    |         expected `i32`, found `&{integer}`
    |         help: consider dereferencing the borrow: `*b`
 
-error: aborting due to 11 previous errors
+error[E0308]: mismatched types
+  --> $DIR/deref-suggestion.rs:63:9
+   |
+LL |         b
+   |         ^
+   |         |
+   |         expected `i32`, found `&{integer}`
+   |         help: consider dereferencing the borrow: `*b`
+
+error: aborting due to 12 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

